design_notes/XDN05 - xlang Binary Interface.md

```diff
@@ -12,6 +12,233 @@ status: draft
+[Platform Abstraction Layer](XDN06%20-%20xlang%20Platform%20Abstraction%20Layer.md).
+
+## COM Interfaces
+
+> This section is intended to be a short introduction to COM interfaces.
+For a more thorough definition of COM interfaces, please review the official Microsoft
+[COM Technical Overview](https://docs.microsoft.com/en-us/windows/desktop/com/com-technical-overview).
+
+[COM](https://docs.microsoft.com/en-us/windows/desktop/com/component-object-model--com--portal)
+is a Microsoft technology that defines a binary interoperability standard for creating reusable
+software libraries that interact at run time. xlang leverages
+[COM interfaces](https://docs.microsoft.com/en-us/windows/desktop/com/com-objects-and-interfaces)
+as well as the definition of fundamental COM interface
+[IUnknown](https://docs.microsoft.com/en-us/windows/desktop/com/using-and-implementing-iunknown)
+as a foundation for stable contracts that both components and language projections can depend on.
+
+A COM interface pointer is basically a pointer to an array of function pointers, similar to how
+most C++ compilers implement
+[virtual method tables](https://en.wikipedia.org/wiki/Virtual_method_table).
+Each member of a COM interface is stored in an array slot as a function pointer. Each function in
+the array takes the array pointer as it's first parameter. While this design is intrinsic to most
+C++ compilers, these function pointer arrays can also be built up explicitly in other languages
+where needed.
+
+COM interfaces are identified by a 128-bit integer known as a globally unique identifier or
+[GUID](https://en.wikipedia.org/wiki/Universally_unique_identifier).
+This identifier enables calling code to query a given xlang component about its support for
+specific interfaces.
+
+Once a COM interface is published, it cannot be changed. As stated before, a COM interface is a
+stable contract between caller and callee. Interoperability would be impossible if COM interfaces
+were allowed to change over time.
+
+COM Interfaces are allowed to inherit from other interfaces. This is often used to add capabilities
+to later versions of a given component. By inheriting from another interface, the COM interface is
+defining a single longer array of function pointers, where the first set of function pointers are
+defined by the base interface while the next set of function pointers are defined by the derived
+interface. Multiple levels of COM interface single inheritance is supported.
+
+> Note, while xlang uses COM interfaces as a stable contract between components, xlang is
+explicitly avoiding the rest of the COM standard, including concepts such as apartments, monikers
+and marshalling. These constructs are tightly bound to the Windows platform and so have little use
+in the xlang project. Likewise, xlang is explicitly avoiding Windows technologies that build on COM
+such as DCOM, MTS, COM+, ActiveX and the Windows Runtime. Windows Runtime is an inspiration for
```

Comment:
Such statements continue to confuse/concern developers. While the xlang project defines a new PAL/ABI that is distinct from the Windows Runtime, the xlang project also provides libraries and tools that target and support the Windows Runtime.

Reply:
The xlang project in the large, sure. But is that true of the xlang abi? 

I want to avoid folks thinking we're porting over a bunch of Windows COM infrastructure to non-windows platforms. Can you suggest alternative phrasing?